src/Workspaces/CoreTestUtilities/Remote/TestSerializerService.cs

```diff
@@ -64,9 +69,82 @@ public override MetadataReference ReadMetadataReferenceFrom(ObjectReader reader,
+            }
+        }
+
+        public override AnalyzerReference ReadAnalyzerReferenceFrom(ObjectReader reader, CancellationToken cancellationToken)
+        {
+            var testGeneratorReferenceGuid = reader.ReadGuid();
+
+            if (testGeneratorReferenceGuid != Guid.Empty)
+            {
+                Contract.ThrowIfFalse(_sharedTestGeneratorReferences.TryGetValue(testGeneratorReferenceGuid, out var generatorReference));
+                return generatorReference;
+            }
+            else
+            {
+                return base.ReadAnalyzerReferenceFrom(reader, cancellationToken);
+            }
+        }
+
         [ExportWorkspaceServiceFactory(typeof(ISerializerService), layer: ServiceLayer.Test), Shared, PartNotDiscoverable]
+        [Export(typeof(Factory))]
         internal new sealed class Factory : IWorkspaceServiceFactory
         {
+            private ConcurrentDictionary<Guid, TestGeneratorReference> _sharedTestGeneratorReferences;
+
+            /// <summary>
+            /// Gate to serialize reads/writes to <see cref="_sharedTestGeneratorReferences"/>.
+            /// </summary>
+            private readonly object _gate = new object();
+
+            /// <summary>
+            /// In unit tests that are testing OOP, we want to be able to share test generator references directly
+            /// from one side to another. We'll create multiple instances of the TestSerializerService -- one for each
+            /// workspace and in different MEF compositions, but we'll share the generator references across them. This
+            /// property allows the shared dictionary to be read/set.
+            /// </summary>
+            public ConcurrentDictionary<Guid, TestGeneratorReference> SharedTestGeneratorReferences
+            {
+                get
+                {
+                    lock (_gate)
+                    {
+                        if (_sharedTestGeneratorReferences == null)
+                            _sharedTestGeneratorReferences = new ConcurrentDictionary<Guid, TestGeneratorReference>();
+
+                        return _sharedTestGeneratorReferences;
+                    }
+                }
+
+                set
+                {
+                    lock (_gate)
+                    {
+                        // If we're already being assigned the same set of references as before, we're fine as that won't change anything.
+                        // Ideally, every time we created a new RemoteWorkspace we'd have a new MEF container; this would ensure that
+                        // the assignment earlier before we create the RemoteWorkspace was always the first assignment. However the
+                        // ExportProviderCache.cs in our unit tests hands out the same MEF container multpile times instead of implementing the expected
+                        // contract. See https://github.com/dotnet/roslyn/issues/25863 for further details.
```

Comment:
good to know!

Reply:
> Ideally, every time we created a new RemoteWorkspace we'd have a new MEF container

This isn't how RemoteWorkspace should work. The whole system is designed so one MEF container can support multiple workspaces simultaneously. If you want one instance of a service associated with one workspace, use `IWorkspaceServiceFactory` and it will do exactly that.

Reply:
@sharwell: agreed, there's no reason they can't be shared. But this isn't the only place that needs documentation changed. The fact TestComposition.GetHostServices() can return a shared instance despite using an IExportProviderFactory that's _supposed_ to create new compositions made it appear to me that the intent of the system is we _should_ have separate compositions and https://github.com/dotnet/roslyn/issues/25863 is tracking the fact it's not. Debugging this was not fun. ☹️

Reply:
I'll see if I can strengthen the requirements there.